### ISSUE TITLE
Tensors have different dimensions

### ISSUE THREAD

**Issue by brchlluk** [ayooshkathuria/pytorch-yolo-v3#1]
Hi,
firstly, thank you for your work on this repo. I tried to run your code, but I get this exception:
```
Traceback (most recent call last):
  File "/home/lukas/dev/pytorch-yolo-v3/detect.py", line 183, in <module>
    prediction = write_results(prediction, confidence, num_classes, nms = True, nms_conf = nms_thesh)
  File "/home/lukas/dev/pytorch-yolo-v3/util.py", line 189, in write_results
    output = torch.cat(seq, 0)
RuntimeError: invalid argument 0: Tensors must have same number of dimensions: got 1 and 2 at /tmp/pip-w41ywlv_-build/aten/src/THC/generic/THCTensorMath.cu:102

```
I tried to debug what is happening in the code, but it is not very clear to me. I only noticed that in my case, one tensor's size is 7 and the other is (7,1). I am also adding a screen.
![image](https://user-images.githubusercontent.com/18639856/38768333-6d9dd7c4-3ff2-11e8-84e2-3caa5d2dec71.png)
 
So I have modified the 7 sized tensor into (7,1) dimension but then I get another exception from further code.
```
Traceback (most recent call last):
  File "/opt/pycharm-2018.1/helpers/pydev/pydevd.py", line 1664, in <module>
    main()
  File "/opt/pycharm-2018.1/helpers/pydev/pydevd.py", line 1658, in main
    globals = debugger.run(setup['file'], None, None, is_module)
  File "/opt/pycharm-2018.1/helpers/pydev/pydevd.py", line 1068, in run
    pydev_imports.execfile(file, globals, locals)  # execute the script
  File "/opt/pycharm-2018.1/helpers/pydev/_pydev_imps/_pydev_execfile.py", line 18, in execfile
    exec(compile(contents+"\n", file, 'exec'), glob, loc)
  File "/home/lukas/dev/pytorch-yolo-v3/detect.py", line 213, in <module>
    objs = [classes[int(x[-1])] for x in output if int(x[0]) == im_id]
  File "/home/lukas/dev/pytorch-yolo-v3/detect.py", line 213, in <listcomp>
    objs = [classes[int(x[-1])] for x in output if int(x[0]) == im_id]
IndexError: list index out of range
```

Don't you please know what I am doing wrong? It is my first time working with PyTorch, so I am not very experienced and I don't know how to fix it myself.

Using Ubuntu 16.4, Python 3.6, CUDA 9.0, PyTorch 0.4.
Thanks

**Comment by brchlluk** [comment]
Perfect, thank your for your quick response. I think this is caused by PyTorch 0.4 version. With your explanation, it is more clear to me now and I will try to fix it in free time. Will make pull request if I get successful.

**Comment by brchlluk** [comment]
Well, I did something similar I saw in your code 
`image_pred_class = image_pred_class.new(image_pred_class.size(0), 1).fill_(image_pred_class[0])`. 
Yea, I know this is not the best way, but it worked :D 

Finally, I found the issue when I checked your structure you wrote and my structure in "prediction" variable in detect.py. The problem is that PyTorch now handles the tensors little differently in my opinion. I had the values of box coordinates and class ids written in single column likes this:
```
   0.0000   59.8087
   0.0000   95.7915
   0.0000  313.1047
   0.0000  308.8701
   0.0000    0.9959
   0.0000    0.9978
   0.0000    1.0000
   0.0000  256.9860
   0.0000   63.2367
   0.0000  374.7674
   0.0000  120.6163
   0.0000    0.9986
   0.0000    0.8848
   0.0000    7.0000
   0.0000   67.9070
   0.0000  164.1937
   0.0000  174.7524
   0.0000  386.2670
   0.0000    0.9999
   0.0000    0.9997
   0.0000   16.0000
[torch.cuda.FloatTensor of size (21,2) (GPU 0)]
```
When I add a little dirty hack on line 200 in util.py, it now works on 0.4 version perfectly.
`image_pred_class = image_pred_class.unsqueeze(1)
 image_pred_class = torch.transpose(image_pred_class, 0, 1)`

I am not making pull request because this would break your current solution and it isn't elegant either. I think someone with more experience could fix it better. 
Thank you for your help!

**Comment by ayooshkathuria** [comment]
Okay. What exactly have you printed above? the (21,2) tensor? Downloading 0.4 now.

**Comment by brchlluk** [comment]
Sorry for not being clear. The above (21,2) tensor is is output from method "write_results", which is called at line 183 in detect.py. It is also content of the variable named "prediction". 
I run the detection only with single image (dog-cycle-car.png) which has 3 detectable objects in it. This explains the (21,2) structure and the error "list index out of range" I mentioned before. 

All this can be fixed just by transposing the tensor in util.py. I only added  
```
image_pred_class = image_pred_class.unsqueeze(1)
image_pred_class = torch.transpose(image_pred_class, 0, 1)
```
on line 200 in util.py and it started working.

**Comment by brchlluk** [comment]
Ok, so I tried on the entire folder and found out that this hotfix doesn't work. The reason is that if there are multiple objects of same class in one image the variable "image_pred_class" is not just (7,)  but it turns into (X,7) where X is the number of the same objects in image. So the unsqueeze would give it another dimension and turn it into (1,X,7) which is obviously wrong.

Solution to this would be to make image_pred_class two dimensional independent to number of same classes in image. This is now broken due to PyTorch 0.4 version incompatibility.

To sum it up. The image_pred_class is one dimensonal column vector (should be row vector), when there is only one object of a kind in image and is 2d matrix (which is correct) when there are more objects of same kind. So this has to be fixed, to run it on 0.4 version, but I don't know how. Maybe someone more experienced with the same problem and 0.4 version will look into it.

**Comment by ayooshkathuria** [comment]
Okay. Thanks for your effort. I'm closing this issue as of now. I'll try to solve it when I get my hands on 0.4 this week. However, I guess it's okay to defer this issue at least until 0.4 becomes the channel that is available in conda/pip channels since a lot of people get their PyTorch from their. Meanwhile, I'll update the readme. I think I'll focus on ability to train the model on COCO before I'll get back to this issue.

**Comment by ayooshkathuria** [comment]
Can you, btw, check whether the folder version works normally on PyTorch 0.4 (without hotfix). Stayed up all night trying to compile PyTorch 0.4 on macOS, and it wouldn't go through :(

**Comment by brchlluk** [comment]
No, it won't run on the folder, because there are images containing only one same class but also there is image when there are multiple horses. So it will crash with or without hotfix.
